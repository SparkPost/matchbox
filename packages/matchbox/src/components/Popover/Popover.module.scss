@import '../../styles/config';

.Popover {
  display: block;
  position: absolute;
  top: 100%;
  left: 0;
  min-width: rem(200);

  background: color(gray, 10);
  margin: 13px 0 0;

  box-shadow: shadow(deep);
  border-radius: border-radius(large);
  border: 1px solid color(gray, 8);

  opacity: 0;
  pointer-events: none;
  transition: 0.1s ease-out;
  transition-property: opacity;

  z-index: 1;

  &.entered {
    pointer-events: auto;
    opacity: 1;
  }

  &.exiting, &.exited {
    transition: 0s;
    opacity: 0;
  }

  &.exited {
    pointer-events: none;
  }

  &.sectioned {
    padding: spacing();
  }
}

.Tip {
  display: block;
  position: absolute;
  top: -13px;
  width: 13px;
  height: 13px;

  margin: 6px 0 0;

  background: color(gray, 10);
  border-top-left-radius: border-radius(large);
  border-top: 1px solid color(gray, 8);
  border-left: 1px solid color(gray, 8);
  transform: translate(-50%, 0) rotate(45deg);

  z-index: 1;
}

.Content {
  position: relative;
  z-index: 1;
  border-radius: border-radius(large);
}

.Activator {
  position: relative;
}

.top {
  .Tip {
    top: auto;
    bottom: -13px;
    margin: 0 0 6px;

    border-top: none;
    border-left: none;
    border-bottom: 1px solid color(gray, 8);
    border-right: 1px solid color(gray, 8);
    border-bottom-right-radius: border-radius(large);
  }

  .Popover {
    top: auto;
    bottom: 100%;
    margin: 0 0 13px;
  }
}

.left {
  .Popover {
    left: auto;
    right: 0;
  }

  .Tip {
    left: auto;
    transform: translate(50%, 0) rotate(45deg);
  }
}
