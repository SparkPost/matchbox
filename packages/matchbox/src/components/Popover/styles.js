import { tokens } from '@sparkpost/design-tokens';

export const content = props => `
  position: absolute;
  background: ${tokens.color_white};
  border: ${props.theme.borders['300']};
  border-radius: ${tokens.borderRadius_100};
  box-shadow: ${tokens.boxShadow_100};
  margin-top: ${props.isTop ? tokens.spacing_0 : tokens.spacing_100};
  margin-bottom: ${props.isTop ? tokens.spacing_100 : tokens.spacing_0};

  top: ${props.isTop ? 'auto' : '100%'};
  bottom: ${!props.isTop ? 'auto' : '100%'};
`;

export const transition = props => {
  const visible = `
    opacity: 1;
    transform: translateY(0);
    transition: opacity ${tokens.motionDuration_medium} ${tokens.motionEase_out},
                transform ${tokens.motionDuration_medium} ${tokens.motionEase_out};
  `;

  const hidden = `
    opacity: 0;
    pointer-events: none;
    transform: translateY(${props.isTop ? tokens.spacing_100 : `-${tokens.spacing_100}`});
    transition: opacity ${tokens.motionDuration_fast} ${tokens.motionEase_in},
                transform ${tokens.motionDuration_fast} ${tokens.motionEase_in};
  `;

  switch (props.state) {
    case 'entered':
      return `
        ${visible}
        pointer-events: auto;
        visibility: visible;
      `;
    case 'exiting':
      return `
        ${hidden}
        visibility: visible;
      `;
    case 'exited':
    case 'entering':
    default:
      return `
        ${hidden}
        visibility: hidden;
      `;
  }
};
