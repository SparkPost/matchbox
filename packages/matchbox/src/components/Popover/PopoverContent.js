import React from 'react';
import { Transition } from 'react-transition-group';
import classnames from 'classnames';
import styles from './Popover.module.scss';

const Content = ({
  children,
  popoverRef,
  open,
  activatorWidth,
  top, left, bottom, right, fixed,
  sectioned,
  className = '',
  trigger,
  ...rest
}) => {

  const popoverClasses = classnames(
    styles.Popover,
    sectioned && styles.sectioned,
    className
  );

  const wrapperClasses = classnames(
    styles.Wrapper,
    top && styles.top,
    left && styles.left
  );

  const tipStyle = { [left ? 'right' : 'left']: activatorWidth / 2 };

  return (
    <Transition mountOnEnter unmountOnExit in={open} timeout={{ enter: 0, exit: 0 }}>
      {(state) => (
        <div className={wrapperClasses} ref={popoverRef}>
          <div className={classnames(popoverClasses, state && styles[state])} {...rest}>
            <span className={styles.Tip} style={tipStyle} />
            <div className={styles.Content}>
              {children}
            </div>
          </div>
        </div>
      )}
    </Transition>
  );
};

Content.displayName = 'Popover.Content';
export default Content;
