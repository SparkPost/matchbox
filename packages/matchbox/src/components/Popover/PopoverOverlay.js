import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Portal } from '../Portal';
import { WindowEvent } from '../WindowEvent';
import { getPositionFor } from '../../helpers/geometry';
import classnames from 'classnames';

import styles from './PopoverOverlay.module.scss';

const defaultPosition = {
  top: 0,
  left: 0,
  width: 0,
  height: 0
};

class PopoverOverlay extends Component {
  static displayName = 'PopoverOverlay';

  static propTypes = {
    renderActivator: PropTypes.func.isRequired,
    renderPopover: PropTypes.func.isRequired
  }

  state = {
    position: defaultPosition
  }

  componentDidMount() {
    this.handleMeasurement();
  }

  componentDidUpdate({ open: prevOpen }) {
    if (prevOpen !== this.props.open) {
      this.handleMeasurement();
    }
  }

  handleMeasurement = () => {
    if (!this.props.open) {
      this.setState({ position: defaultPosition });
    } else {
      this.setState({
        position: getPositionFor(this.activator, { fixed: this.props.fixed })
      });
    }
  }

  render() {
    const { renderPopover, renderActivator, portalId, fixed, open } = this.props;
    const { position } = this.state;

    const activatorProps = {
      activatorRef: (node) => this.activator = node
    };

    const overlayClasses = classnames(styles.PopoverOverlay, fixed && styles.fixed);

    return (
      <Fragment>
        {open && <WindowEvent event='resize' handler={this.handleMeasurement} />}
        {(!fixed && open) ? <WindowEvent event='scroll' handler={this.handleMeasurement} /> : null}
        {renderActivator(activatorProps)}
        <Portal containerId={portalId}>
          <div className={overlayClasses} style={position}>
            {renderPopover({ activatorWidth: this.state.position.width })}
          </div>
        </Portal>
      </Fragment>
    );
  }
}

export default PopoverOverlay;
