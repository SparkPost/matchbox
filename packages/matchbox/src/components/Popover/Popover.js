import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { WindowEvent } from '../WindowEvent';
import PopoverOverlay from './PopoverOverlay';
import PopoverContent from './PopoverContent';
import { onKey } from '../../helpers/keyEvents';

import styles from './Popover.module.scss';

class Popover extends Component {
  static displayName = 'Popover';

  static propTypes = {
    /**
     * A React component to will trigger the popover
     * Click event is handled for you if this component is uncontrolled.
     */
    trigger: PropTypes.node,
    /**
      * Adds a padding to the Popover
      */
    sectioned: PropTypes.bool,
    /**
      * Opens the popover.
      * By default, open state is handled automatically. Passing this value in will turn this into a controlled component.
      */
    open: PropTypes.bool,
    left: PropTypes.bool,
    right: PropTypes.bool,
    top: PropTypes.bool,
    bottom: PropTypes.bool,
    /**
      * Callback function that is called when clicking outside the popover, or hitting escape.
      */
    onClose: PropTypes.func,
    /**
      * Popover Content
      */
    children: PropTypes.node,
    /**
     * Element ID for the portal that will house Popover. Appends to body if not provided.
     */
    portalId: PropTypes.string
  };

  static defaultProps = {
    right: true,
    bottom: true
  }

  state = {
    open: null
  }

  componentDidMount() {
    const { open: controlledOpen } = this.props;

    // This component becomes "controlled" if the prop 'open' is given a boolean value
    if (controlledOpen === undefined) {
      this.setState({ open: false });
    }
  }

  uncontrolledToggle = () => {
    this.setState({ open: !this.state.open });
  }

  handleOutsideClick = (e) => {
    const { open: controlledOpen, onClose } = this.props;
    const isOutside = this.popover && !this.popover.contains(e.target) && this.activator && !this.activator.contains(e.target);

    if (isOutside && (controlledOpen || this.state.open)) {
      if (onClose) {
        onClose(e);
      }

      if (this.state.open) {
        this.uncontrolledToggle();
      }
    }
  }

  handleEsc = (e) => {
    const { open: controlledOpen, onClose } = this.props;

    if (onClose && (controlledOpen || this.state.open)) {
      onKey('escape', onClose)(e);
    }

    if (this.state.open) {
      onKey('escape', this.uncontrolledToggle)(e);
    }
  }

  handleTrigger = () => {
    if (this.state.open === false) {
      this.uncontrolledToggle();
    }
  }

  renderPopover = ({ activatorWidth }) => {
    const {
      children,
      open: controlledOpen,
      ...rest
    } = this.props;

    const shouldBeOpen = controlledOpen || this.state.open;

    return (
      <PopoverContent
        open={shouldBeOpen}
        popoverRef={(node) => this.popover = node}
        activatorWidth={activatorWidth}
        {...rest}>
        <WindowEvent event='click' handler={this.handleOutsideClick} />
        <WindowEvent event='keydown' handler={this.handleEsc} />
        {children}
      </PopoverContent>
    );
  }

  renderActivator = ({ activatorRef }) => {
    const {
      trigger,
      wrapper: WrapperComponent = 'span'
    } = this.props;

    const assignRefs = (node) => {
      this.activator = node;
      activatorRef(node);
    };

    return (
      <WrapperComponent
        className={styles.Activator}
        onClick={this.handleTrigger}
        ref={assignRefs}>
        {trigger}
      </WrapperComponent>
    );
  }

  render() {
    const { fixed, open: controlledOpen } = this.props;
    const shouldBeOpen = controlledOpen || this.state.open;

    return (
      <PopoverOverlay
        fixed={fixed}
        open={shouldBeOpen}
        portalId={this.props.portalId}
        renderActivator={this.renderActivator}
        renderPopover={this.renderPopover} />
    );
  }
}

export default Popover;
