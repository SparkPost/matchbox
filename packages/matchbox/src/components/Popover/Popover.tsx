import React from 'react';
import { PaddingProps, LayoutProps, ResponsiveValue } from 'styled-system';
import { PortalProps } from '../Portal';
import { Box } from '../Box';
import PopoverOverlay from './PopoverOverlay';
import PopoverContent from './PopoverContent';
import { onKey, onKeys } from '../../helpers/keyEvents';
import useWindowEvent from '../../hooks/useWindowEvent';
import { findFocusableChild } from '../../helpers/focus';
import type * as Polymorphic from '../../helpers/types';
import { combineRefs } from '../../helpers/ref';

export type PopoverProps = PaddingProps &
  LayoutProps & {
    id?: string;
    /**
     * A React component to will trigger the popover
     * Click event is handled for you if this component is uncontrolled.
     */
    trigger?: React.ReactNode;
    /**
     * Adds a padding to the Popover
     * @deprecated Use padding system props instead
     */
    sectioned?: boolean;
    /**
     * Opens the popover.
     * By default, open state is handled automatically. Passing this value in will turn this into a controlled component.
     */
    open?: boolean;
    /**
     * @deprecated Use `position` instead
     */
    left?: boolean;
    /**
     * @deprecated Use `position` instead
     */
    right?: boolean;
    /**
     * @deprecated Use `position` instead
     */
    top?: boolean;
    /**
     * @deprecated Use `position` instead
     */
    bottom?: boolean;
    position?: ResponsiveValue<'topLeft' | 'topRight' | 'bottomLeft' | 'bottomRight'>;
    /**
     * Callback function that is called when clicking outside the popover, or hitting escape.
     */
    onClose?: (e: MouseEvent | KeyboardEvent) => void;

    /**
     * Popover Content
     */
    children?: React.ReactNode;
    as?: 'div' | 'span';
    /**
     * @deprecated Use `as` prop instead
     */
    wrapper?: 'div' | 'span';
    portalId?: PortalProps['containerId'];
    /**
     * Closes the popover when tabbing out of the popover. On by default.
     */
    closeOnTab?: boolean;
    /**
     * Closes the popover when clicking inside the popover. Off by default.
     */
    closeOnInsideClick?: boolean;
  };

type PolymorphicPopover = Polymorphic.ForwardRefComponent<'span', PopoverProps>;

const Popover = React.forwardRef<HTMLSpanElement, PopoverProps>(function Popover(props, ref) {
  const {
    as,
    id,
    open: controlledOpen,
    onClose,
    children,
    trigger,
    wrapper,
    portalId,
    closeOnTab = true,
    closeOnInsideClick = false,
    ...rest
  } = props;
  const [open, setOpen] = React.useState(null);
  const popoverRef = React.useRef<HTMLDivElement>();
  const activatorRef = React.useRef<HTMLDivElement>();

  const shouldBeOpen = controlledOpen || open;
  const Wrapper = as || wrapper || 'span';

  useWindowEvent('click', handleClick);
  useWindowEvent('keydown', handleEsc);

  // Mount hook â€“ sets initial close state if uncontrolled
  React.useEffect(() => {
    // This component becomes "controlled" if the prop 'open' is given a boolean value
    if (controlledOpen === undefined) {
      setOpen(false);
    }
  }, []);

  // Automatically focuses on content when opening
  React.useEffect(() => {
    if (shouldBeOpen && popoverRef && popoverRef.current) {
      const contentToFocus = findFocusableChild(popoverRef.current) || popoverRef.current;

      // Honestly not sure why this doesn't work without a timeout
      setTimeout(() => {
        contentToFocus.focus();
      }, 10);
    }
  }, [shouldBeOpen]);

  // Handles `aria-haspopup` and `aria-expanded` attributes on the activator
  React.useEffect(() => {
    if (activatorRef && activatorRef.current) {
      const activatorElem = findFocusableChild(activatorRef.current) || activatorRef.current;
      activatorElem.setAttribute('aria-haspopup', 'true');
      activatorElem.setAttribute('aria-expanded', Boolean(shouldBeOpen).toString());
    }
  }, [trigger, activatorRef, open, controlledOpen]);

  // Toggles uncontrolled open state
  function handleUncontrolledToggle() {
    setOpen(!open);
  }

  // Focus on activator element
  // This is only called for when closing via keyboard
  function focusOnActivator() {
    if (activatorRef && activatorRef.current) {
      const activatorToFocus = findFocusableChild(activatorRef.current) || activatorRef.current;
      activatorToFocus.focus();
    }
  }

  // Toggles uncontrolled popovers on clicking outside, and calls `onClose` for controlled popovers
  function handleClick(e: MouseEvent) {
    const isInside = popoverRef.current && popoverRef.current.contains(e.target as Node);
    const isOutside =
      popoverRef.current &&
      !popoverRef.current.contains(e.target as Node) &&
      activatorRef.current &&
      !activatorRef.current.contains(e.target as Node);

    if ((isOutside && shouldBeOpen) || (isInside && closeOnInsideClick && shouldBeOpen)) {
      if (onClose) {
        onClose(e);
      }

      if (open) {
        handleUncontrolledToggle();
      }
    }
  }

  // Toggles uncontrolled popovers on escape keydown, and calls `onClose` for controlled popovers
  function handleEsc(e: KeyboardEvent) {
    if (onClose && shouldBeOpen) {
      onKey('escape', () => {
        onClose(e);
        focusOnActivator();
      })(e);
    }

    if (onClose && shouldBeOpen && closeOnTab) {
      onKey('tab', () => {
        onClose(e);
        focusOnActivator();
      })(e);
    }

    // Uncontrolled
    if (open) {
      onKey('escape', () => {
        handleUncontrolledToggle();
        focusOnActivator();
      })(e);
    }

    // Uncontrolled
    if (closeOnTab && open) {
      onKey('tab', () => {
        handleUncontrolledToggle();
        focusOnActivator();
      })(e);
    }
  }

  // Toggles uncontrolled popovers
  function handleTrigger() {
    if (open !== null) {
      handleUncontrolledToggle();
    }
  }

  function handleActivatorKey(e: React.KeyboardEvent) {
    if (open === false) {
      onKeys(['arrowUp', 'arrowDown'], () => {
        // Stop arrow keys from scrolling the page
        e.preventDefault();
        handleUncontrolledToggle();
      })(e);
    }
  }

  // Renders popover content
  function renderPopover() {
    return (
      <PopoverContent open={shouldBeOpen} ref={combineRefs(ref, popoverRef)} {...rest}>
        {children}
      </PopoverContent>
    );
  }

  // Renders popover trigger
  function renderActivator({ activatorRef: forwardedRef }) {
    return (
      <Box
        as={Wrapper}
        // Inline block is required to measure and set height correctly
        display={Wrapper === 'span' ? 'inline-block' : null}
        position="relative"
        onClick={handleTrigger}
        onKeyDown={handleActivatorKey}
        ref={combineRefs(forwardedRef, activatorRef)}
      >
        {trigger}
      </Box>
    );
  }

  return (
    <PopoverOverlay
      as={Wrapper}
      id={id}
      open={shouldBeOpen}
      renderActivator={renderActivator}
      renderPopover={renderPopover}
      portalId={portalId}
    />
  );
}) as PolymorphicPopover & {
  PopoverContent: typeof PopoverContent;
  PopoverOverlay: typeof PopoverOverlay;
};

Popover.displayName = 'Popover';

export default Popover;
