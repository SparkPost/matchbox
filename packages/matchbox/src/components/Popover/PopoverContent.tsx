import React from 'react';
import { Transition } from 'react-transition-group';
import styled from 'styled-components';
import { padding, PaddingProps, layout, LayoutProps, compose, variant } from 'styled-system';
import { pick } from '@styled-system/props';
import { Box } from '../Box';
import { tokens } from '@sparkpost/design-tokens';
import { secondsToMS } from '../../helpers/string';
import { content, transition } from './styles';
import Popover from './Popover';

const system = compose(layout, padding);

const StyledContent = styled('div')`
  ${system}
  ${content}
  ${transition}
  ${variant({
    variants: {
      bottomLeft: {
        right: '0',
        left: 'auto',
        top: '100%',
        bottom: 'auto',
        mt: '100',
      },
      bottomRight: {
        left: '0',
        right: 'auto',
        top: '100%',
        bottom: 'auto',
        mt: '100',
      },
      topRight: {
        left: '0',
        right: 'auto',
        top: 'auto',
        bottom: '100%',
        mb: '100',
      },
      topLeft: {
        right: '0',
        left: 'auto',
        top: 'auto',
        bottom: '100%',
        mb: '100',
      },
    },
  })}
`;

function getPositionFromDeprecatedProps({ top, left }: { top?: boolean; left?: boolean }) {
  if (left && !top) {
    return 'bottomLeft';
  }

  if (!left && !top) {
    return 'bottomRight';
  }

  if (left && top) {
    return 'topLeft';
  }

  if (!left && top) {
    return 'topRight';
  }
}

type ContentProps = PaddingProps &
  LayoutProps &
  Pick<
    React.ComponentProps<typeof Popover>,
    'children' | 'top' | 'left' | 'position' | 'className' | 'sectioned' | 'open'
  > & {
    style?: React.CSSProperties;
  };

const Content = React.forwardRef<HTMLDivElement, ContentProps>(function Content(
  props: ContentProps,
  userRef,
) {
  const {
    children,
    open,

    // TODO: to be removed
    top = false,
    left = false,

    // TODO: default to bottomRight when other deprecated props are removed
    position,
    sectioned,
    className = '',
    style,
    ...rest
  } = props;

  const transitionRef = React.useRef();
  const systemProps = pick(rest);

  // TODO: Remove in future
  const variant = position || getPositionFromDeprecatedProps({ top, left });

  return (
    <Transition
      mountOnEnter
      unmountOnExit
      in={open}
      timeout={{
        enter: 0,
        exit: secondsToMS(tokens.motionDuration_fast),
      }}
      nodeRef={transitionRef}
    >
      {(state) => (
        <Box position="relative" height="100%" ref={userRef} tabIndex={-1}>
          <StyledContent
            data-id="popover-content"
            className={className}
            p={sectioned ? '400' : null}
            minWidth="13rem" // 208px
            variant={variant}
            // TODO: remove in future, isTop is still needed for animation
            isTop={variant.toString().includes('top')}
            state={state}
            {...style}
            {...systemProps}
            ref={transitionRef}
          >
            <div>{children}</div>
          </StyledContent>
        </Box>
      )}
    </Transition>
  );
});

Content.displayName = 'Popover.Content';
export default Content;
