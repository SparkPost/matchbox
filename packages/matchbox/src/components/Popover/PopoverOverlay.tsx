import React from 'react';
import { WindowEvent } from '../WindowEvent';
import { Portal } from '../Portal';
import { Box } from '../Box';
import { getPositionFor } from '../../helpers/geometry';
import { tokens } from '@sparkpost/design-tokens';
import Popover from './Popover';

const defaultPosition = {
  top: 0,
  left: 0,
  width: 0,
  height: 0,
};

type OverlayProps = Pick<React.ComponentProps<typeof Popover>, 'id' | 'open' | 'portalId'> & {
  renderPopover: () => React.ReactNode;
  renderActivator: (any) => React.ReactNode;
  as?: 'div' | 'span';
};

function PopoverOverlay(props: OverlayProps): JSX.Element {
  const [position, setPosition] = React.useState(defaultPosition);
  const activatorRef = React.useRef(null);
  const { as, id, open, renderPopover, renderActivator, portalId } = props;

  function handleMeasurement() {
    setPosition(getPositionFor(activatorRef.current));
  }

  React.useEffect(() => {
    handleMeasurement();
  }, [open]);

  return (
    <>
      <WindowEvent event="resize" handler={handleMeasurement} />
      <Box
        as={as}
        // Inline block is required to measure and set height correctly on spans
        display={as === 'span' ? 'inline-block' : null}
        position="relative"
      >
        {renderActivator({
          activatorRef: (node) => {
            activatorRef.current = node;
          },
        })}
        <Portal containerId={portalId}>
          <Box
            {...(!open ? { 'aria-hidden': true } : {})}
            id={id}
            position="absolute"
            left={`${position.left}px`}
            top={`${position.top}px`}
            height={`${position.height}px`}
            width={`${position.width}px`}
            zIndex={tokens.zIndex_overlay + 1}
            style={{ pointerEvents: 'none' }}
          >
            {renderPopover()}
          </Box>
        </Portal>
      </Box>
    </>
  );
}

PopoverOverlay.displayName = 'PopoverOverlay';

export default PopoverOverlay;
