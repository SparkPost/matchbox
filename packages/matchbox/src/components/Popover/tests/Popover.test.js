import React from 'react';
import Popover from '../Popover';
import { shallow } from 'enzyme';
import * as keyEventHelpers from '../../../helpers/keyEvents';
import * as geometryHelpers from '../../../helpers/geometry';

describe('Popover', () => {
  const Trigger = (props) => <div {...props} />;
  let wrapper;
  let instance;
  let activator;
  let popover;
  const activatorRefMock = jest.fn();
  let toggleSpy;

  beforeEach(() => {
    wrapper = shallow(<Popover trigger={<Trigger/>} portalId='portal'>Popover Content</Popover>);
    instance = wrapper.instance();
    activator = () => shallow(wrapper.instance().renderActivator({ activatorRef: activatorRefMock }));
    popover = () => shallow(wrapper.instance().renderPopover({ activatorWidth: 100 }));
    keyEventHelpers.onKey = jest.fn(() => jest.fn());
    geometryHelpers.getPositionFor = jest.fn(() => 'mock position');
    toggleSpy = jest.spyOn(wrapper.instance(), 'uncontrolledToggle');
  });

  it('should render overlay correctly', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('PopoverOverlay').dive()).toMatchSnapshot();
  });

  it('should render window events when open', () => {
    wrapper.setProps({ open: true });
    expect(wrapper.find('PopoverOverlay').dive().find('WindowEvent')).toHaveLength(4);
  });

  it('should measure position when opened and revert to default when closed', () => {
    wrapper.setProps({ open: true });
    expect(wrapper.find('PopoverOverlay').dive().find('.PopoverOverlay').prop('style')).toMatchSnapshot();
    wrapper.setProps({ open: false });
    expect(wrapper.find('PopoverOverlay').dive().find('.PopoverOverlay').prop('style')).toMatchSnapshot();
  });

  it('should use local state if not controlled', () => {
    expect(shallow(<Popover />)).toHaveState({ open: false });
  });

  it('should use not use local state if a control prop is provided', () => {
    expect(shallow(<Popover open={false} />)).toHaveState({ open: null });
  });

  describe('activator render', () => {
    it('should render correctly', () => {
      expect(activator()).toMatchSnapshot();
    });

    it('should render a wrapper component correctly', () => {
      wrapper.setProps({ wrapper: 'div' });
      expect(activator()).toMatchSnapshot();
    });

    it('should handle an uncontrolled open', () => {
      expect(wrapper.state().open).toBe(false);
      activator().simulate('click');
      expect(wrapper.state().open).toBe(true);
    });
  });

  describe('popover render', () => {
    it('should render correctly', () => {
      expect(popover()).toMatchSnapshot();
    });

    it('should be open if controlled or uncontrolled', () => {
      expect(popover().props().in).toEqual(false);
      wrapper.setProps({ open: true });
      expect(popover().props().in).toEqual(true);
      wrapper.setProps({ open: false });
      activator().simulate('click');
      expect(popover().props().in).toEqual(true);
    });

    it('should use correct classnames', () => {
      wrapper.setProps({ sectioned: true, top: true, left: true });
      expect(popover().props().children()).toMatchSnapshot();
    });
  });

  describe('window events', () => {

    describe('escape', () => {
      it('should do nothing if popover is not open', () => {
        wrapper.setProps({ onClose: jest.fn() });
        wrapper.instance().handleEsc('event');
        expect(keyEventHelpers.onKey).not.toHaveBeenCalled();
      });

      it('should call onClose prop if controlled component', () => {
        wrapper.setProps({ open: true, onClose: jest.fn() });
        wrapper.instance().handleEsc('event');
        expect(keyEventHelpers.onKey).toHaveBeenCalledWith('escape', instance.props.onClose);
        expect(keyEventHelpers.onKey).not.toHaveBeenCalledWith('escape', wrapper.instance().uncontrolledToggle);
      });

      it('should close popover if uncontrolled', () => {
        wrapper.setProps({ onClose: jest.fn() });
        wrapper.instance().setState({ open: true });
        wrapper.instance().handleEsc('event');
        expect(keyEventHelpers.onKey).toHaveBeenCalledWith('escape', instance.props.onClose);
        expect(keyEventHelpers.onKey).toHaveBeenCalledWith('escape', wrapper.instance().uncontrolledToggle);
      });
    });

    describe('outside click', () => {
      describe('is outside', () => {

        beforeEach(() => {
          wrapper.setProps({ onClose: jest.fn() });
          instance.popover = { contains: jest.fn(() => false) };
          instance.activator = { contains: jest.fn(() => false) };
        });

        it('should close if controlled open', () => {
          wrapper.setProps({ open: true });
          instance.handleOutsideClick('event');
          expect(toggleSpy).not.toHaveBeenCalled();
          expect(instance.props.onClose).toHaveBeenCalled();
        });

        it('should close if uncontrolled open', () => {
          instance.setState({ open: true });
          instance.handleOutsideClick('event');
          expect(toggleSpy).toHaveBeenCalled();
          expect(instance.props.onClose).toHaveBeenCalled();
        });

        it('should not close if closed', () => {
          instance.setState({ open: false });
          instance.handleOutsideClick('event');
          wrapper.setProps({ open: false });
          instance.handleOutsideClick('event');
          expect(toggleSpy).not.toHaveBeenCalled();
          expect(instance.props.onClose).not.toHaveBeenCalled();
        });
      });

      describe('is inside', () => {
        beforeEach(() => {
          wrapper.setProps({ onClose: jest.fn() });
          instance.popover = { contains: jest.fn(() => true) };
          instance.activator = { contains: jest.fn(() => true) };
        });

        it('should not close', () => {
          instance.setState({ open: true });
          instance.handleOutsideClick('event');
          expect(toggleSpy).not.toHaveBeenCalled();
          expect(instance.props.onClose).not.toHaveBeenCalled();
        });
      });
    });
  });

  describe('handle trigger toggle', () => {
    it('should do nothing if popover is controlled', () => {
      instance.setState({ open: null }); // Reverts to default state
      instance.handleTrigger();
      expect(toggleSpy).not.toHaveBeenCalled();
    });

    it('should open popover if uncontrolled', () => {
      instance.handleTrigger();
      expect(toggleSpy).toHaveBeenCalled();
    });
  });
});
