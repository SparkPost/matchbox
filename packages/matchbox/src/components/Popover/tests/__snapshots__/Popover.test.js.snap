// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover activator render should render a wrapper component correctly 1`] = `
<div
  className="Activator"
  onClick={[Function]}
>
  <Trigger />
</div>
`;

exports[`Popover activator render should render correctly 1`] = `
<span
  className="Activator"
  onClick={[Function]}
>
  <Trigger />
</span>
`;

exports[`Popover popover render should render correctly 1`] = `
<Transition
  appear={false}
  enter={true}
  exit={true}
  in={false}
  mountOnEnter={true}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={
    Object {
      "enter": 0,
      "exit": 0,
    }
  }
  unmountOnExit={true}
>
  <Component />
</Transition>
`;

exports[`Popover popover render should use correct classnames 1`] = `
<div
  className="Wrapper top left"
>
  <div
    className="Popover sectioned"
    portalId="portal"
  >
    <span
      className="Tip"
      style={
        Object {
          "right": 50,
        }
      }
    />
    <div
      className="Content"
    >
      <WindowEvent
        event="click"
        handler={[Function]}
      />
      <WindowEvent
        event="keydown"
        handler={[Function]}
      />
      Popover Content
    </div>
  </div>
</div>
`;

exports[`Popover should measure position when opened and revert to default when closed 1`] = `"mock position"`;

exports[`Popover should measure position when opened and revert to default when closed 2`] = `
Object {
  "height": 0,
  "left": 0,
  "top": 0,
  "width": 0,
}
`;

exports[`Popover should render overlay correctly 1`] = `
<PopoverOverlay
  open={false}
  portalId="portal"
  renderActivator={[Function]}
  renderPopover={[Function]}
/>
`;

exports[`Popover should render overlay correctly 2`] = `
<Fragment>
  <span
    className="Activator"
    onClick={[Function]}
  >
    <Trigger />
  </span>
  <Portal
    containerId="portal"
  >
    <div
      className="PopoverOverlay"
      style={
        Object {
          "height": 0,
          "left": 0,
          "top": 0,
          "width": 0,
        }
      }
    >
      <Popover.Content
        activatorWidth={0}
        bottom={true}
        open={false}
        popoverRef={[Function]}
        portalId="portal"
        right={true}
        trigger={<Trigger />}
      >
        <WindowEvent
          event="click"
          handler={[Function]}
        />
        <WindowEvent
          event="keydown"
          handler={[Function]}
        />
        Popover Content
      </Popover.Content>
    </div>
  </Portal>
</Fragment>
`;
